File: agent.js
const { runLLMAnalysis } = require("./llm");
const {
  closeClient,
  getFile,
  getRepoConfig,
  initializeClient,
  searchCode
} = require("./mcpClient");

async function fetchCodeContext(searchQueries) {
  const context = [];
  const seenPaths = new Set();
  const observations = [];

  for (const query of searchQueries) {
    const matches = await searchCode(query);
    observations.push({
      type: "search",
      query,
      matches: matches.slice(0, 5)
    });

    for (const match of matches) {
      if (seenPaths.has(match.path)) {
        continue;
      }

      const file = await getFile(match.path);
      context.push(file);
      observations.push({
        type: "file",
        path: file.path,
        preview: String(file.content).slice(0, 500)
      });
      seenPaths.add(match.path);
    }
  }

  return {
    context,
    observations
  };
}

async function executeRepositoryTool(name, args, state) {
  await initializeClient();

  if (name === "search_code") {
    const matches = await searchCode(args.query);
    state.observations.push({
      type: "search",
      query: args.query,
      matches: matches.slice(0, 5)
    });
    state.mcp.attempted = true;
    state.mcp.used = true;
    state.mcp.searches += 1;
    state.mcp.tool_observations += 1;

    return {
      matches: matches.slice(0, 10)
    };
  }

  if (name === "get_file_contents") {
    if (state.seenPaths.has(args.path)) {
      const existingFile = state.codeContext.find((file) => file.path === args.path);

      return {
        path: args.path,
        sha: existingFile?.sha || null,
        content: existingFile?.content || ""
      };
    }

    const file = await getFile(args.path);
    state.codeContext.push(file);
    state.seenPaths.add(file.path);
    state.observations.push({
      type: "file",
      path: file.path,
      preview: String(file.content).slice(0, 500)
    });
    state.mcp.attempted = true;
    state.mcp.used = true;
    state.mcp.files_loaded += 1;
    state.mcp.tool_observations += 1;

    return file;
  }

  throw new Error(`Unsupported repository tool: ${name}`);
}

async function runAgent(ticket) {
  const repository = getRepoConfig();
  const mcp = {
    attempted: false,
    used: false,
    searches: 0,
    files_loaded: 0,
    tool_observations: 0
  };

  let codeContext = [];
  let observations = [];
  let ticketAnalysis;
  let technicalAnalysis;
  const toolState = {
    codeContext,
    observations,
    mcp,
    seenPaths: new Set()
  };

  try {
    const result = await runLLMAnalysis({
      ticket,
      repository,
      executeTool: async (name, args) => executeRepositoryTool(name, args, toolState),
      fallbackContext: {
        codeContext,
        observations
      }
    });

    ticketAnalysis = result.ticket_analysis;
    const executedQueries = observations
      .filter((item) => item.type === "search")
      .map((item) => item.query);
    if (executedQueries.length > 0) {
      ticketAnalysis.search_queries = executedQueries;
    }
    technicalAnalysis = result.technical_analysis;
  } finally {
    await closeClient();
  }

  return {
    repository,
    ticket,
    ticket_analysis: ticketAnalysis,
    mcp,
    observations,
    code_context: codeContext,
    technical_analysis: technicalAnalysis
  };
}

module.exports = {
  runAgent,
  fetchCodeContext
};

File: agent.py
from typing import Any

from agent_config import get_agent_config
from llm import run_llm_analysis
from mcp_client import GitHubMcpClient, get_repo_config


def is_useful_code_path(path: str) -> bool:
    normalized = path.replace("\\", "/").lower()
    noisy_segments = ("/obj/", "/bin/", "/.git/", "/packages/")
    noisy_suffixes = (".g.cs", ".designer.cs", ".user", ".cache")

    if any(segment in normalized for segment in noisy_segments):
        return False

    if normalized.endswith(noisy_suffixes):
        return False

    return True


async def prefetch_repository_context(
    search_queries: list[str],
    state: dict[str, Any],
    client: GitHubMcpClient,
    max_files: int,
) -> int:
    await client.initialize()
    loaded_count = 0

    for query in search_queries:
        if len(state["code_context"]) >= max_files:
            return loaded_count

        matches = await client.search_code(query)
        state["observations"].append(
            {
                "type": "search",
                "query": query,
                "matches": matches[:5],
            }
        )

        for match in matches:
            if len(state["code_context"]) >= max_files:
                return loaded_count

            path = match["path"]
            if not is_useful_code_path(path):
                continue
            if path in state["seen_paths"]:
                continue

            file = await client.get_file(path)
            state["code_context"].append(file)
            state["seen_paths"].add(file["path"])
            state["observations"].append(
                {
                    "type": "file",
                    "path": file["path"],
                    "preview": str(file["content"])[:500],
                }
            )
            loaded_count += 1

    return loaded_count


async def prefetch_prioritized_files(
    paths: list[str],
    state: dict[str, Any],
    client: GitHubMcpClient,
    max_files: int,
) -> int:
    await client.initialize()
    loaded_count = 0

    for path in paths:
        if len(state["code_context"]) >= max_files:
            return loaded_count
        if not is_useful_code_path(path):
            continue
        if path in state["seen_paths"]:
            continue

        file = await client.get_file(path)
        state["code_context"].append(file)
        state["seen_paths"].add(file["path"])
        state["observations"].append(
            {
                "type": "file",
                "path": file["path"],
                "preview": str(file["content"])[:500],
            }
        )
        loaded_count += 1

    return loaded_count


async def run_agent(ticket: str) -> dict[str, Any]:
    config = get_agent_config()
    repository = get_repo_config()
    code_context: list[dict[str, Any]] = []
    observations: list[dict[str, Any]] = []
    state = {
        "code_context": code_context,
        "observations": observations,
        "seen_paths": set(),
    }
    client = GitHubMcpClient()
    ticket_analysis: dict[str, Any] = {
        "problem_type": None,
        "needs_code_context": True,
        "suspected_areas": [],
        "planner_notes": "",
        "search_queries": [],
    }

    try:
        result: dict[str, Any] | None = None

        for round_index in range(1, config.policy.max_analysis_rounds + 1):
            result = await run_llm_analysis(
                ticket=ticket,
                repository=repository,
                observations=observations,
                code_context=code_context,
                round_index=round_index,
                max_rounds=config.policy.max_analysis_rounds,
            )
            ticket_analysis = result["ticket_analysis"]
            technical_analysis = result["technical_analysis"]

            confidence_score = int(technical_analysis.get("confidence_score") or 0)
            reached_threshold = confidence_score >= config.policy.confidence_threshold
            reached_limit = round_index >= config.policy.max_analysis_rounds
            needs_more_context = bool(technical_analysis.get("needs_more_context"))

            if reached_threshold or reached_limit or not needs_more_context:
                break

            loaded_count = 0
            prioritized_files = technical_analysis.get("prioritized_files") or []
            additional_search_queries = technical_analysis.get("additional_search_queries") or []

            if prioritized_files:
                loaded_count += await prefetch_prioritized_files(
                    prioritized_files,
                    state,
                    client,
                    max_files=config.policy.max_context_files,
                )

            if additional_search_queries and len(code_context) < config.policy.max_context_files:
                loaded_count += await prefetch_repository_context(
                    additional_search_queries,
                    state,
                    client,
                    max_files=config.policy.max_context_files,
                )

            if loaded_count == 0:
                technical_analysis["needs_more_context"] = False
                break

        if result is None:
            raise RuntimeError("Ticket analysis did not produce a result.")

        ticket_analysis["search_queries"] = [
            item["query"] for item in observations if item.get("type") == "search"
        ]

        return {
            "repository": repository,
            "ticket": ticket,
            "ticket_analysis": ticket_analysis,
            "observations": observations,
            "code_context": code_context,
            "technical_analysis": result["technical_analysis"],
        }
    finally:
        await client.close()
